### PR TITLE
Add Event Study (aka Dynamic Difference in Differences) functionality

### LINKED ISSUE #384: Event Study and Staggered Diff in Diff
Working on this issue now. Will aim to have a PR after the major code refactor. Currently just exploring modelling options...

### PR DESCRIPTION
**Pull request by drbenvincent**
Closes #384

This PR adds **Event Study / Dynamic DiD** for **simultaneous treatment timing** (TWFE + event-time dummies). The staggered-adoption half of #384 landed on `main` in #621 (`StaggeredDifferenceInDifferences`); this PR is the complementary simultaneous-timing estimator and explicitly rejects staggered adoption.

## Summary

- New `EventStudy` experiment class (PyMC + sklearn), with `effect_summary()`, unified `plot()` API, and `generate_event_study_data()` simulator
- Docs: `event-study-pymc.ipynb`, gallery entry, glossary, reporting statistics table
- Cross-cutting NaN handling: `build_formula_matrices(..., return_type="dataframe")` + row filtering across several experiment classes (bugbot finding)
- Synced with current `main` (gallery.yaml, formula utils, PyMC 6 notebook conventions, etc.)

## PR checklist

- [x] This implementation does not deal with staggered treatments. Add a warning callout box, as data validation with exception
- [x] Update model description - the model being used is not static, it depends on the user provided formula. Give some examples.
- [x] Add an example where we remove the time fixed effects and add in other predictors or some form of seasonal trend.
- [x] Check the integration with the reporting layer.
- [ ] ~~Consider adding the time relative to treatment column in the notebook, not hidden in the experiment class.~~
- [x] Ensure we are exposing plot options via arguments
- [x] Simplify data plot using a 1-liner with seaborn
- [x] Why is the event study column not appearing in the table on home page?

## Since Dec review

- Merged latest `main` and resolved conflicts
- Added method-level tests (`test_event_study.py`, `test_nan_handling.py`)
- Addressed bugbot items: `np.inf` control units, `round_to` in `get_event_time_summary`, NaN row/index alignment
- CI green on latest head

## Deferred (follow-up issues, not blocking this PR)

- Real-world marketing datasets for the notebook
- Demeaning / alternative FE handling for very large unit counts
- Patsy `Treatment(reference=...)` refactor for event-time dummies

---

⚠️ Includes a more widespread bug fix identified by bugbot here https://github.com/pymc-labs/CausalPy/pull/584#discussion_r2595103978 which affects multiple experiment classes.

----
📚 Documentation preview 📚: https://causalpy--584.org.readthedocs.build/en/584/

### CODE REVIEW

**Comment by juanitorduz** [comment]
I will take a look in the next few days :)

**Comment by nialloulton** [comment]
@cursor review

File: causalpy/experiments/event_study.py [hunk 1]
```diff
@@ -0,0 +1,648 @@
...
+    ...     model=cp.pymc_models.LinearRegression(
+    ...         sample_kwargs={
+    ...             "tune": 20,
+    ...             "draws": 20,
+    ...             "chains": 2,
+    ...             "progressbar": False,
+    ...             "random_seed": 42,
+    ...         }
+    ...     ),
+    ... )  # doctest: +SKIP
```
Comment (juanitorduz) [inline causalpy/experiments/event_study.py:135]: Can we add some references? For example https://arxiv.org/pdf/2503.13323

File: causalpy/experiments/event_study.py [hunk 2]
```diff
@@ -0,0 +1,636 @@
...
+            raise DataException(
+                "Data contains duplicate unit-time observations. "
+                "Each unit should have at most one observation per time period."
+            )
+
+        # Check that all treated units have the same treatment time
+        # (staggered adoption is not currently supported)
+        treated_times = self.data.loc[
+            ~self.data[self.treat_time_col].isna(), self.treat_time_col
+        ].unique()
```
[the remark this answers is not in the document]
Comment (juanitorduz) [reply causalpy/experiments/event_study.py]: we should fix this inconsistency and add a unit-test ;)

File: causalpy/experiments/event_study.py [hunk 3]
```diff
@@ -0,0 +1,636 @@
...
+        self._y_design_info = y.design_info
+        self._x_design_info = X.design_info
+
+        # Extract outcome variable name from formula
+        self.outcome_variable_name = y.design_info.column_names[0]
+
+        # Convert patsy output to DataFrames for manipulation
+        X_df = pd.DataFrame(
+            X, columns=X.design_info.column_names, index=self.data.index
+        )
```
[the remark this answers is not in the document]
Comment (juanitorduz) [reply causalpy/experiments/event_study.py]: we should fix this inconsistency and add a unit-test ;)

File: causalpy/experiments/event_study.py [hunk 4]
```diff
@@ -0,0 +1,648 @@
...
+        event_time_dummies = pd.DataFrame(index=self.data.index)
+        for k in event_times_non_ref:
+            col_name = f"event_time_{k}"
+            # 1 if treated and at event time k, 0 otherwise
+            event_time_dummies[col_name] = (
+                (self.data["_event_time"] == k) & self.data["_in_event_window"]
+            ).astype(float)
+
+        # Combine patsy design matrix with event-time dummies
+        X_df = pd.concat([X_df, event_time_dummies], axis=1)
```
Comment (juanitorduz) [inline causalpy/experiments/event_study.py:336]: @drbenvincent, if the number of units gets very big, then fitting all these dummies will become hard. What  about adding the option to "demean" the data instead of using https://github.com/py-econometrics/pyfixest ? We can still use Bayesian methods to do inference after the demeaning procedure (just an idea)

File: causalpy/experiments/event_study.py [hunk 5]
```diff
@@ -0,0 +1,636 @@
...
+                    "event_time": k,
+                    "mean": float(coeff),
+                    "std": np.nan,
+                    hdi_lower_col: np.nan,
+                    hdi_upper_col: np.nan,
+                    "is_reference": False,
+                }
+            rows.append(row)
+
+        return pd.DataFrame(rows)
```
[the remark this answers is not in the document]
Comment (juanitorduz) [reply causalpy/experiments/event_study.py:522]: +1

File: causalpy/tests/test_event_study.py [hunk 6]
```diff
@@ -0,0 +1,542 @@
...
+            formula="y ~ C(unit) + C(time)",
+            unit_col="unit",
+            time_col="time",
+            treat_time_col="treat_time",
+            model=cp.pymc_models.LinearRegression(sample_kwargs=sample_kwargs),
+        )
+
+
+# ============================================================================
+# Integration Tests with PyMC
```
Comment (juanitorduz) [inline causalpy/tests/test_event_study.py]: @drbenvincent, can we add a cursor to add tests for each individual method of the `EventStudy` class? This could be especially helpful to detect edge cases (say nans and stuff like that)

**Review by juanitorduz — CHANGES_REQUESTED** [review]
Very cool! This is super helpful in marketing (the effect of a marketing campaign depends on time!). We should try to find other real datasets to showcase this.

Minor comments. 

Suggestion -> add more tests for the indidual methods and address @bugbot comments

File: causalpy/experiments/event_study.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (drbenvincent) [reply causalpy/experiments/event_study.py]: Resolved by https://github.com/pymc-labs/CausalPy/pull/584/commits/76986a2056af369c5745037e7ab9ae388d496d8b

File: causalpy/experiments/event_study.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (drbenvincent) [reply causalpy/experiments/event_study.py:522]: Should be resolved by https://github.com/pymc-labs/CausalPy/pull/584/commits/364faed3b43222af8552d4507afbe531d921cd7a I think

File: causalpy/experiments/event_study.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (drbenvincent) [reply causalpy/experiments/event_study.py]: resoled by https://github.com/pymc-labs/CausalPy/pull/584/commits/964903d6823649d207310ba0cade89eefd5f8a12

File: causalpy/tests/test_event_study.py (hunk 6 above)
Comment (drbenvincent) [reply causalpy/tests/test_event_study.py]: Added a bunch of tests in https://github.com/pymc-labs/CausalPy/pull/584/commits/02294ec46f41e514fb9039382a82df702fb03b70

File: causalpy/experiments/event_study.py (hunk 4 above)
Comment (drbenvincent) [reply causalpy/experiments/event_study.py:336]: Good idea, but a) not sure if we need an expternal package for de-meaning? b) let's have this as an iteration on top of the first mvp (i.e. yes, but not in this PR)

File: causalpy/experiments/event_study.py (hunk 1 above)
Comment (drbenvincent) [reply causalpy/experiments/event_study.py:135]: Added some citations in https://github.com/pymc-labs/CausalPy/pull/584/commits/b4eb1bfb32cf91b61801f7c48baf1efafc5714b7

**Comment by drbenvincent** [comment]
<p>Done, along with some other citations</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/CausalPy/pull/584/discussion/#comment-3655260502'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

File: causalpy/experiments/event_study.py (hunk 4 above)
Comment (NathanielF) [reply causalpy/experiments/event_study.py:336]: We could also consider mundlak variations which is just the groupby mean per group. 

```
sim_df['unit_mean'] = sim_df.groupby('unit')['treat'].transform(np.mean)

sim_df['state_mean'] = sim_df.groupby('state')['treat'].transform(np.mean)

sim_df['cohort_mean'] = sim_df.groupby('cohort_year')['treat'].transform(np.mean)

sim_df['time_mean'] = sim_df.groupby('year')['treat'].transform(np.mean)

```

File: causalpy/experiments/event_study.py [hunk 7]
```diff
@@ -0,0 +1,714 @@
...
+    Example
+    --------
+    >>> import causalpy as cp
+    >>> from causalpy.data.simulate_data import generate_event_study_data
+    >>> df = generate_event_study_data(
+    ...     n_units=20, n_time=20, treatment_time=10, seed=42
+    ... )
+    >>> result = cp.EventStudy(
+    ...     df,
+    ...     formula="y ~ C(unit) + C(time)",
```
Comment (NathanielF) [inline causalpy/experiments/event_study.py:120]: I think you should be able to encode the formula fully in patsy. Something like 

"y ~ C(unit) + C(rel_time, Treatment(reference=-1))"

It just requires that you pre-processe the df['time'] column before running patsy. 

Something like 

df = df.sort_values(["unit", "time"])

df["t_idx"] = df.groupby("unit").cumcount()

df["event_idx"] = (
    df[df["treated"] == 1]
    .groupby("unit")["t_idx"]
    .transform("min")
)

df["rel_time"] = df["t_idx"] - df["event_idx"]

Comment (drbenvincent) [reply causalpy/experiments/event_study.py:120]: Sorry for the AI evaluation, but some major concerns with that proposal

--- 

## The Suggestion

Use patsy's categorical handling to simplify the event-time dummy creation:

```python
"y ~ C(unit) + C(rel_time, Treatment(reference=-1))"
```

With pre-processing to compute `rel_time` before calling patsy.

## Benefits

- Simpler code (removes ~50 lines of manual dummy creation)
- More idiomatic use of patsy
- Reference category handled automatically via `Treatment(reference=-1)`

## Concerns

### 1. Control Units Would Be Dropped

This is the most significant issue. Control units have `NaN` for treatment time, so `rel_time` would be `NaN`. Patsy drops rows with `NaN` values entirely.

**Current behavior**: Control units are kept in the estimation. Their event-time dummies are all 0, but they contribute to the time fixed effects estimation.

**Proposed behavior**: Control units would be excluded, fundamentally changing the model and breaking the parallel trends comparison.

### 2. Loss of Calendar Time Fixed Effects

The proposed formula replaces calendar time FEs with event-time dummies. But a proper TWFE event study needs **both**:

```
Y_it = α_i + λ_t + Σ β_k · 1{E_it = k} + ε_it
        ↑      ↑         ↑
      unit   time    event-time
```

With only `C(rel_time)`, we lose `λ_t` (calendar time effects). This changes the identifying assumptions of the model.

### 3. Event Window Handling

Currently, only event times within the window (e.g., -5 to +5) get dummies. Observations outside the window are kept but don't contribute to event-time indicators.

With patsy handling everything, we'd need to either:
- Filter out observations with `rel_time` outside the window (loses data)
- Let patsy create dummies for ALL unique `rel_time` values (different model, many more parameters)
- Bin/clip values outside the window to special categories

### 4. Minor Issues

- **Coefficient naming**: Changes from `event_time_0` to `C(rel_time, Treatment(reference=-1))[T.0]`
- **Dynamic reference**: Would need to build formula strings dynamically to inject `reference_event_time`
- **Test updates**: Existing tests check current coefficient naming

## Possible Middle Ground

Keep the current model structure (which correctly handles control units and maintains both time FEs and event-time dummies) but simplify the internal implementation where possible.

## Questions

1. Am I missing something that would address the control unit issue?
2. Is the loss of calendar time FEs intentional in your suggestion, or should both be included?
3. Any thoughts on handling the event window bounds?

Happy to implement if we can resolve these concerns!

Comment (drbenvincent) [reply causalpy/experiments/event_study.py:120]: ## Initial Concerns

The naive approach has some issues:

1. **Control units dropped**: NaN `rel_time` causes patsy to drop entire rows
2. **Loss of calendar time FEs**: Replacing `C(time)` with `C(rel_time)` loses λ_t
3. **Event window handling**: How to handle observations outside the window?

## The Elegant Solution

**Key insight**: In patsy's `Treatment` coding, the reference category gets 0 for all dummies. This is exactly what we want for:
- The reference period (k=-1)
- Control units
- Observations outside the event window

**Solution: Map all these cases to the reference category!**

```python
# Step 1: Compute relative time as usual
df["rel_time"] = df["time"] - df["treat_time"]

# Step 2: Create a categorical version where:
#   - Control units → reference_event_time
#   - Outside event window → reference_event_time
#   - Inside window → actual rel_time
df["rel_time_cat"] = df["rel_time"].copy()

# Control units (NaN rel_time) get the reference category
df.loc[df["rel_time"].isna(), "rel_time_cat"] = reference_event_time

# Observations outside window also get reference category
outside_window = (df["rel_time"] < event_window[0]) | (df["rel_time"] > event_window[1])
df.loc[outside_window & df["rel_time"].notna(), "rel_time_cat"] = reference_event_time

# Step 3: Use patsy with BOTH time FEs and event-time dummies
formula = f"y ~ C(unit) + C(time) + C(rel_time_cat, Treatment(reference={reference_event_time}))"
```

## Why This Works

| Observation Type | `rel_time_cat` Value | Result |
|------------------|---------------------|--------|
| Treated, k=-1 (reference) | -1 | 0 for all event dummies ✓ |
| Treated, k in window | actual k | Gets its own dummy ✓ |
| Treated, k outside window | -1 (reference) | 0 for all event dummies ✓ |
| Control unit | -1 (reference) | 0 for all event dummies ✓ |

## Benefits

1. **Calendar time FEs preserved** - `C(time)` stays in the formula
2. **Control units handled correctly** - They contribute to FE estimation, get 0 for event dummies
3. **Event window handled naturally** - Outside-window observations map to reference
4. **Much simpler code** - Patsy creates all the dummies
5. **No rows dropped** - No NaN values in the categorical column

## Implementation Approach

The user could still provide a simple formula, and the class would augment it internally:

```python
# User provides:
formula = "y ~ C(unit) + C(time)"

# Class internally transforms to:
augmented_formula = f"y ~ C(unit) + C(time) + C(rel_time_cat, Treatment(reference={ref}))"
```

This significantly simplifies `_build_design_matrix()` while preserving all the correct model behavior.

Comment (drbenvincent) [reply causalpy/experiments/event_study.py:120]: What do you think @NathanielF? Maybe we merge as is, create an issue for further work, then come back to it in a second iteration?

Comment (NathanielF) [reply causalpy/experiments/event_study.py:120]: This makes sense.

**Review by NathanielF — APPROVED** [review]
Sorry it took so long.

**Comment by drbenvincent** [comment]
## Automated triage

**Recommendation:** `review:high` — no decision gate identified.

**Why:**
- Major new feature: `EventStudy` class (759 lines), new `reporting.py` logic (201 lines), new data simulation functions.
- Cross-cutting: changes to `base.py`, `reporting.py`, multiple experiment files, glossary, and references.
- Large test suite: `test_event_study.py` (1087 lines), `test_nan_handling.py`, `test_reporting.py`, `test_synthetic_data.py`.
- All CI checks pass. PR is conflicted (DIRTY). Review already has requested changes. Includes a bug fix affecting multiple experiment classes.

**Review focus:**
1. Verify the event study model handles non-staggered treatment correctly (the PR notes this limitation with a warning).
2. Review the bug fix that affects multiple experiment classes (noted in the PR description).
3. Confirm the reporting layer integration works for event study-specific outputs.

**Confidence:** high

**Comment by drbenvincent** [comment]
@juanitorduz — when you have a moment, could you take another look at this PR on the current head (`95bbfc5c`)?

Your Dec review asked for more per-method tests and bugbot fixes. Since then we've:

- merged latest `main` and resolved conflicts
- added extensive `EventStudy` tests plus `test_nan_handling.py` across experiment classes
- fixed the bugbot items (`np.inf` controls, `round_to` in `get_event_time_summary`, NaN row/index alignment via `build_formula_matrices` + index filtering)
- added unified `plot()` on `EventStudy` and gallery notebook `event-study-pymc`

CI is green on the latest push. Deferred items from the thread (real marketing datasets, demeaning FE for large N, patsy `Treatment()` refactor) are noted in the updated PR description as follow-ups.

If you're happy with the current state, an approval would unblock merge. If anything still feels like a major sticking point, let us know and we'll address it.
